glib/src/subclass/type_module.rs

```diff
@@ -0,0 +1,145 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use crate::{subclass::prelude::*, translate::*, Cast, IsA, TypeModule};
+
+pub trait TypeModuleObserver<T: IsA<TypeModule>> {
+    fn on_load(type_module: &T) -> bool;
+    fn on_unload(type_module: &T);
+}
+
+pub trait TypeModuleImpl: ObjectImpl + TypeModuleImplExt {
+    // rustdoc-stripper-ignore-next
+    /// Loads the module, registers one or more types using
+    /// [`register_module_type`] and registers one or more interfaces using
+    /// [`register_module_interface`] (see [`TypeModule`]).
+    ///
+    /// [`register_module_type`]: ../types/fn.register_module_type.html
+    /// [`register_module_interface`]: ../interface/fn.register_module_interface.html
+    /// [`TypeModule`]: ../../gobject/auto/type_module/struct.TypeModule.html
+    fn load(&self) -> bool;
+    // rustdoc-stripper-ignore-next
+    /// Unloads the module (see [`TypeModuleExt::unuse`]).
+    ///
+    /// [`TypeModuleExt::unuse`]: ../../gobject/auto/type_module/trait.TypeModuleExt.html#method.unuse
+    fn unload(&self);
+}
+
+pub trait TypeModuleImplExt: ObjectSubclass {
+    fn parent_load(&self) -> bool;
+    fn parent_unload(&self);
+}
+
+impl<T: TypeModuleImpl> TypeModuleImplExt for T {
+    fn parent_load(&self) -> bool {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().parent_class() as *const gobject_ffi::GTypeModuleClass;
+
+            let f = (*parent_class)
+                .load
+                .expect("No parent class implementation for \"load\"");
+
+            from_glib(f(self
+                .obj()
+                .unsafe_cast_ref::<TypeModule>()
+                .to_glib_none()
+                .0))
+        }
+    }
+
+    fn parent_unload(&self) {
+        unsafe {
+            let data = T::type_data();
+            let parent_class = data.as_ref().parent_class() as *const gobject_ffi::GTypeModuleClass;
+
+            let f = (*parent_class)
+                .unload
+                .expect("No parent class implementation for \"unload\"");
+
+            f(self.obj().unsafe_cast_ref::<TypeModule>().to_glib_none().0);
+        }
+    }
+}
+
+unsafe impl<T: TypeModuleImpl> IsSubclassable<T> for TypeModule {
+    fn class_init(class: &mut crate::Class<Self>) {
+        Self::parent_class_init::<T>(class);
+
+        let klass = class.as_mut();
+        klass.load = Some(load::<T>);
+        klass.unload = Some(unload::<T>);
+    }
+}
+
+unsafe extern "C" fn load<T: TypeModuleImpl>(
+    type_module: *mut gobject_ffi::GTypeModule,
+) -> ffi::gboolean {
+    let instance = &*(type_module as *mut T::Instance);
+    let imp = instance.imp();
+
+    let res = imp.load();
+
+    if res {
+        unsafe {
+            gobject_ffi::g_object_ref(type_module as _);
```

Comment:
This should probably get some comment. Why do we ref here (and unref below)?

Reply:
Panicking if zero sounds like a good approach if that solves the problem here.

With the ref/unref as before, wouldn't this cause the type module to never be disposed?

Reply:
So that means we should also just leak the instance here ourselves or not? You could simply call `g_object_ref()` once and never unref it again, that seems simpler IMHO.

Reply:
That sounds reasonable